### ISSUE TITLE
irb and async?

### ISSUE THREAD

**Issue by emiltin** [socketry/async#48]
Hi,
Is it possible to have async tasks run in the background in irb? For example, a server that handles connections in the background, and which you can interact with:

> server = Server.new
=> #<Server...>
> server.start
=> "Server running"
> server.connections
=> 0
> server.connections   # a bit later
=> 6

**Comment by emiltin** [comment]
I guess I can start my Server (and the Async reactor) in a thread. But then I can't use things like waiting for states of the server, which internally use Async::Notification?

**Comment by ioquatix** [comment]
It should be possible to feed non-blocking IO into IRB, but I haven’t tried it.

@aycabta any thoughts? What interface is required to do it?

**Comment by emiltin** [comment]
Do you mean running IRB inside an Async block, and using Async IO objects as input/output to IRB?

**Comment by ioquatix** [comment]
Yes it should be possible to do that. You can check if the IRB call is within an Async task and then use async io stdin/stdout.

**Comment by emiltin** [comment]
Gave it a try, and it seems to work fairly well:
https://gist.github.com/emiltin/d382d1527d771afa8130b174e323539f

```ruby
# counter.rb
require 'async'
class Counter
	attr_reader :count
	def initialize
		@count = 0
		Async do |task|
			loop do
				task.sleep 1
				@count = @count + 1
			end
		end
	end
end
```

```
Emils-MacBook-Air-2:async-irb emiltin$ ./irb
irb(main):001:0> require './counter'
=> true
irb(main):002:0> c=Counter.new
=> #<Counter:0x00007fcac906c750 @count=0>
irb(main):003:0> c.count
=> 2
irb(main):004:0> c.count.    # let's wait a bit and look again
=> 17                        # yay, it's counting in the background!
irb(main):005:0> sleep 20    # will this work?
=> 20         
irb(main):006:0> c.count.    # 20 seconds later
=> 24                        # not right, we need use async task.sleep
```

**Comment by emiltin** [comment]
Looking at https://github.com/ruby/irb/blob/master/lib/irb/input-method.rb, encoding() and eof() probably also needs to be implemented.
Another thing that doesn't work is arrow-up to select previous lines, any idea why that is?

**Comment by emiltin** [comment]
Sleeping can be done with `Async::Task.current.sleep 10`. Would it make sense to have a class method for putting the curent task to sleep, like `Async.sleep`?

**Comment by emiltin** [comment]
When you have async tasks that output stuff in the background, it can be hard to see what you're typing. I was hoping irb sesions could be used to keep noisy tasks in the background, while you interact with them in a separate session.

But sessions are based on threads and background sessions seems to be put to sleep:

```
irb#1(main):009:0> jobs
=> #0->irb on main (#<Thread:0x00007fa697064088 sleep_forever>: stop)
#1->irb#1 on main (#<Thread:0x00007fa697888278@/Users/emiltin/.rbenv/versions/2.6.3/lib/ruby/2.6.0/irb/ext/multi-irb.rb:192 run>: running)
```

**Comment by ioquatix** [comment]
The way I wanted to solve this was by having a separate process connect to the running process from a different terminal.

**Comment by ioquatix** [comment]
@aycabta wondering if you've had time to explore this?

**Comment by ioquatix** [comment]
This is now fixed in Async 2.0.

**Comment by emiltin** [comment]
Goods news!
How is it used? If I just use an Async block, it only returns once it's complete, as expected.

```
irb(main):008:0> t=Async {|task| task.sleep 2}      # blocks for 2 seconds
=> #<Async::Task:0x38720>
irb(main):009:0> t.status
=> :complete
```

How can I run async tasks, while interacting with irb at the same time?

**Comment by emiltin** [comment]
hi @bruno thank you, I just found it too! it basically show want i would like to do, but it seems to be a demo running a customized version of irb? at least i can't reproduce the behaviour with the standard irb in ruby 3.1.0. `Async { binding.irb }` blocks.

**Comment by ioquatix** [comment]
There was an unfortunate bug and this will be fixed in Ruby 3.1.1.
